modules/tinymce/src/core/main/ts/api/html/DomParser.ts

```diff
@@ -332,14 +346,27 @@ const whitespaceCleaner = (root: AstNode, schema: Schema, settings: DomParserSet
       const elementRule = schema.getElementRule(node.name);
       if (validate && elementRule) {
         const isNodeEmpty = isEmpty(schema, nonEmptyElements, whitespaceElements, node);
-        if (elementRule.removeEmpty && isNodeEmpty) {
-          if (blockElements[node.name]) {
-            node.remove();
+
+        const removeOrPadd = () => {
+          if (elementRule.removeEmpty && isNodeEmpty) {
+            if (blockElements[node.name]) {
+              node.remove();
+            } else {
+              node.unwrap();
+            }
+          } else if (elementRule.paddEmpty && (isNodeEmpty || isPaddedWithNbsp(node))) {
+            paddEmptyNode(settings, args, blockElements, node);
+          }
+        };
+
+        if (elementRule.paddInEmptyBlock && isNodeEmpty) {
+          if (isTextRootBlockEmpty(node)) {
+            paddEmptyNode(settings, args, blockElements, node);
           } else {
-            node.unwrap();
+            removeOrPadd();
           }
-        } else if (elementRule.paddEmpty && (isNodeEmpty || isPaddedWithNbsp(node))) {
-          paddEmptyNode(settings, args, blockElements, node);
+        } else {
+          removeOrPadd();
         }
```

Comment:
This is over complicated and can be simplified to something like this I believe:
```ts
        if (elementRule.paddInEmptyBlock && isNodeEmpty && isTextRootBlockEmpty(node)) {
          paddEmptyNode(settings, args, blockElements, node);
        } else if (elementRule.removeEmpty && isNodeEmpty) {
          if (blockElements[node.name]) {
            node.remove();
          } else {
            node.unwrap();
          }
        } else if (elementRule.paddEmpty && (isNodeEmpty || isPaddedWithNbsp(node))) {
          paddEmptyNode(settings, args, blockElements, node);
        }
```
This also helps with readability to make it show the order the properties are evaluated.

Reply:
I had something similar stored in a text file but initially thought an early return would be required. Alhtough, what you have is much better and is equivalent to what I had previously so I will make the change. Thanks

Reply:
Done